Adds RBAC aggregation for the net-attach-def

dougbtv opened the pull request:
Aggregates the viewing and editing for net-attach-defs to users with the view & edit roles.

reported in this BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1721444

danwinship commented on bindata/network/multus/002-rbac.yaml:
```diff
@@ -63,3 +63,48 @@ subjects:
+- apiGroups: ["apiextensions.k8s.io"]
+  resources:
+  - customresourcedefinitions
+  - customresourcedefinitions/status
```
A multus-specific policy should not be giving people permission to view all CustomResourceDefinition objects. If that's actually needed, it should be added to the base role bindings in origin.
(Also, including the `/status` subresource explicitly should not be needed if you're not specifying different permissions from the parent resource.)

But I think this isn't needed anyway; you should be able to give people permission to view net-attach-defs without them also needing permission to view CRDs. I think.

danwinship commented on bindata/network/multus/002-rbac.yaml:
```diff
@@ -63,3 +63,48 @@ subjects:
+metadata:
+  labels: 
+    rbac.authorization.k8s.io/aggregate-to-edit: "true"
+  name: multus:edit
```
This authorization is wrong. We only want cluster administrators to be able to create CRDs and network-attachment-definitions, and they already have that permission.

danwinship commented on bindata/network/multus/002-rbac.yaml:
```diff
@@ -63,3 +63,48 @@ subjects:
+  - list
+  - watch
+- apiGroups: ["k8s.cni.cncf.io"]
+  resources: ["*"]
```
I don't remember all the details of how roles and bindings work, but we need to make sure that this doesn't give project admins the ability to view net-attach-defs in *other* namespaces.

squeed replied:
I needs to be a Role, not a ClusterRole.

squeed commented:
Now that I sit and think for a bit, I feel like ordinary users shouldn't be able to edit or create NetworkAttachmentDefinitions. It is a pretty wide security hole.

dougbtv commented:
Thanks for the input guys, good catches. I've got a new commit pushed up that limits to aggregating to only view permission (that is, it removes the aggregation to the edit permission). Additionally, it removes the overall access to custom resource definitions.

dougbtv commented:
Also fixed up the unit tests to look specifically for the multus:view ClusterRole and also updated the count of objects therein. I made these as incremental commits, please lmk if you'd like me to squash them and rebase, too.

squeed commented:
@dougbtv can you squash? Then I'll lgtm.

dougbtv commented:
Sure thing / No prob @squeed -- have a new-and-rebased single commit up

dougbtv commented:
/test e2e-aws

dougbtv commented:
/retest

squeed commented on bindata/network/multus/002-rbac.yaml:
```diff
@@ -63,3 +63,17 @@ subjects:
   namespace: openshift-multus
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
```
As mentioned below, I believe this needs to be a Role.

dougbtv commented:
Thanks guys for the attention to detail on the Role vs. ClusterRole, edited per input.

squeed commented on bindata/network/multus/002-rbac.yaml:
```diff
@@ -52,14 +52,28 @@ metadata:
 
 ---
 apiVersion: rbac.authorization.k8s.io/v1
-kind: ClusterRoleBinding
```
er, I'm confused. This is the role assigned to the service account used by the multus binary itself, right? Then I doubt you can change this. I believe that adding the new Role below you've done what you need.

dougbtv replied:
Nice catch, this indeed should not have been changed, you're right -- all it needs is the additional Role.

dougbtv commented:
@squeed -- thanks for the poke, new rebased commit up with the changes to not alter the ClusterRole as directly used by Multus.

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/test e2e-aws
/test e2e-aws-upgrade

dougbtv commented:
/retest

dougbtv commented:
/retest

squeed commented:
@dougbtv all the retests in the world aren't going to work:
```
                "conditions": [
                    {
                        "lastTransitionTime": "2019-08-01T21:12:21Z",
                        "message": "Error while updating operator configuration: could not apply (rbac.authorization.k8s.io/v1, Kind=Role) /multus:view: could not create (rbac.authorization.k8s.io/v1, Kind=Role) /multus:view: the server does not allow this method on the requested resource",
                        "reason": "ApplyOperatorConfig",
                        "status": "True",
                        "type": "Degraded"
                    },
```

dougbtv commented:
/retest

squeed commented:
@dougbtv you need to fix the rbac issues. The problem is that you're crating the Role before you're creating the CRD, I think.

dougbtv commented:
@squeed -- thanks for the heads up there, I'm... not sure how I missed that log entry. Thanks for picking that up for me, and for the tip, appreciate it.

dankenigsberg commented:
/cc @booxter is this the cause of https://bugzilla.redhat.com/show_bug.cgi?id=1743985 ?

dougbtv commented:
Pushed a rebase (let's see if the permission are any different for the CNO, now?)

dougbtv commented:
/retest
